### PR TITLE
[API] Add filter categories

### LINKED ISSUE #167: Add extendable filter categories
**Is your feature request related to a problem? Please describe.**
As of now, MLX can't replicate MAX/DAX as fields in the metadata are missing

**Describe the solution you'd like**
Please add the following meta data field to all asset types:
- `metadata` of type [`JSON`](https://dev.mysql.com/doc/refman/5.7/en/json.html#json-paths)

**Describe alternatives you've considered**
Alternatively, metadata for each asset type is static which would make MLX highly inflexible to adjust to the needs of different organizations

### PR DESCRIPTION
**Pull request by ckadner**
**Changes in this PR:**

* Add support for `filter_categories` to all assets **except** `Pipelines`
  ```YAML
  filter_categories:
    domain: "image-to-text"
    platform: 
      - "kubernetes"
      - "kfserving"
    language: "python"
    framework: "tensorflow"
  ```
* The `filter_categories` are represented by a Python `dict` which is stored in MySQL columns of datatype `JSON`
* All `list_...` API endpoints except `pipelines` support filtering by `filter_categories` -- making use of MySQL JSON support
  ```Bash
  curl -X GET --header 'Accept: application/json' 'http://0.0.0.0:8080/apis/v1alpha1/catalog?filter=%7B%22name%22%3A%22%25image%25%22%2C%20%22filter_categories%22%3A%20%7B%22language%22%3A%20%22python%22%2C%20%22platform%22%3A%20%5B%22kfserving%22%2C%20%22kubernetes%22%5D%7D%7D'
  
  # filter='{"name":"%image%", "filter_categories": {"language": "python", "platform": ["kfserving", "kubernetes"]}}'
  ```

**Resolved issues:**
- #167

**Remaining work:**:
- [ ] Add API examples (tests) to show filtering list calls using the new `filter_categories`
- [ ] PR [#16](https://github.com/machine-learning-exchange/katalog/pull/16) on katalog repo to add `filter_categories` to DAX datasets and MAX models

### CODE REVIEW

**Comment by drewbutlerbb4** [comment]
@ckadner Does filter_categories have a set list of options or is it on a "bring your own" categories basis?

> filter_categories are represented by a Python dict

 I do not see a specification so I assume any and all categories are allowed. I think the "bring your own" approach is fine, but I want to get an idea of what the UI might see. Will the API allow categories such as:
```
filter_categories:
  domain:
    platform: 
      - "kubernetes"
      - "kfserving"
```

**Comment by ckadner** [comment]
> @ckadner Does filter_categories have a set list of options or is it on a "bring your own" categories basis?
> 
> > filter_categories are represented by a Python dict
> 
> I do not see a specification so I assume any and all categories are allowed. I think the "bring your own" approach is fine, but I want to get an idea of what the UI might see. Will the API allow categories such as:
> 
> ```
> filter_categories:
>   domain:
>     platform: 
>       - "kubernetes"
>       - "kfserving"
> ```

No, the `dict` must be one-level depth mapping of key-value pairs. The values may be lists however.

I suggest to use the values as labels on the featured cards. The `filter_categories` keys are usefull for grouping the left-side filter checkboxes in the future.

File: api/server/swagger_server/__main__.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ def main():
 
     log.info("MLX API version: %s" % VERSION)
 
-    cx_app = connexion.App(__name__, specification_dir='./swagger/')
+    cx_app = connexion.App("MLX-API", specification_dir='./swagger/')
```
Comment (ckadner) [inline api/server/swagger_server/__main__.py:40]: workaround for debugger problem in PyCharm: 
- https://youtrack.jetbrains.com/issue/PY-49984 "Debugger doesn't work with Flask when called as `Flask(__name__)` - 2021.2 CE" 
- https://youtrack.jetbrains.com/issue/PY-50019 "Error after 2021.2 Update"

Comment (ckadner) [reply api/server/swagger_server/__main__.py:40]: #189 #190

**Pull request merged**